packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt

```diff
@@ -44,18 +50,97 @@ class UpdatesController private constructor(
     UpdatesUtils.getRuntimeVersion(updatesConfiguration)
   )
   val fileDownloader: FileDownloader = FileDownloader(context)
 
+  private val errorRecovery: ErrorRecovery = ErrorRecovery(object : ErrorRecoveryDelegate {
+    override fun loadRemoteUpdate() {
+      if (loaderTask?.isRunning == true) {
+        return
+      }
+      remoteLoadStatus = ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADING
+      val database = getDatabase()
+      val remoteLoader = RemoteLoader(context, updatesConfiguration, database, fileDownloader, updatesDirectory)
+      remoteLoader.start(object : Loader.LoaderCallback {
+        override fun onFailure(e: Exception) {
+          setRemoteLoadStatus(ErrorRecoveryDelegate.RemoteLoadStatus.IDLE)
+          releaseDatabase()
+        }
+        override fun onSuccess(update: UpdateEntity?) {
+          setRemoteLoadStatus(
+            if (update != null) ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADED
+            else ErrorRecoveryDelegate.RemoteLoadStatus.IDLE
+          )
+          releaseDatabase()
+        }
+        override fun onAssetLoaded(asset: AssetEntity, successfulAssetCount: Int, failedAssetCount: Int, totalAssetCount: Int) { }
+        override fun onUpdateManifestLoaded(updateManifest: UpdateManifest) =
+          selectionPolicy.shouldLoadNewUpdate(updateManifest.updateEntity, launchedUpdate, updateManifest.manifestFilters)
+      })
+    }
+
+    override fun relaunch(callback: LauncherCallback) { relaunchReactApplication(context, false, callback) }
+    override fun throwException(exception: Exception) { throw exception }
+
+    override fun markFailedLaunchForLaunchedUpdate() {
+      if (isEmergencyLaunch) {
+        return
+      }
+      databaseHandler.post {
+        val launchedUpdate = launchedUpdate ?: return@post
+        val database = getDatabase()
+        database.updateDao().incrementFailedLaunchCount(launchedUpdate)
+        releaseDatabase()
+      }
+    }
+
+    override fun markSuccessfulLaunchForLaunchedUpdate() {
+      if (isEmergencyLaunch) {
+        return
+      }
+      databaseHandler.post {
+        val launchedUpdate = launchedUpdate ?: return@post
+        val database = getDatabase()
+        database.updateDao().incrementSuccessfulLaunchCount(launchedUpdate)
+        releaseDatabase()
+      }
+    }
+
+    override fun getRemoteLoadStatus() = remoteLoadStatus
+    override fun getCheckAutomaticallyConfiguration() = updatesConfiguration.checkOnLaunch
+    override fun getLaunchedUpdateSuccessfulLaunchCount() = launchedUpdate?.successfulLaunchCount ?: 0
+  })
+
+  private fun setRemoteLoadStatus(status: ErrorRecoveryDelegate.RemoteLoadStatus) {
+    remoteLoadStatus = status
+    errorRecovery.notifyNewRemoteLoadStatus(status)
+  }
+
   // launch conditions
   private var isLoaderTaskFinished = false
   var isEmergencyLaunch = false
     private set
 
+  fun onDidCreateReactInstanceManager(reactInstanceManager: ReactInstanceManager) {
```

Comment:
```suggestion
  fun onCreateReactInstanceManager(reactInstanceManager: ReactInstanceManager) {
```

Reply:
I'd prefer to keep this matching the name of the method in `ReactNativeHostHandler`. (plus I would say that `onDidCreateReactInstanceManager` is slightly more precise, even if less grammatically correct; it's clearer that it's called _after_ the manager has been created.)